### PR TITLE
Initial commit of RabbitMQ support (WIP)

### PR DESCRIPTION
**Pull request by blakepettersson**
This is a work in progress in making RabbitMQ an officially supported database. 

In the container there's an HTTP(S) admin console as well as the broker itself. 

There's a couple of caveats/TODOs though:

* Still need to add BATS tests.
* `USERNAME` and `DATABASE` parameters has to be added.
* I'm not really sure of the best way to handle the certs; RabbitMQ requires that the certificate of the CA is supplied, along with the keypair itself. I added the certs in this PR, but this should probably be handled by supplying the `SSL_CERTIFICATE` and `SSL_KEY ` options, as per the Aptible database spec.
* Not sure if we want to lock down RabbitMQ further; there's a [plethora of ciphers and options](https://www.rabbitmq.com/ssl.html) that can be specified here, but the defaults should be good enough.
* The hostname has to be stable (by using the `-h`option on `docker run`), otherwise you're gonna have a bad time. Not sure if that's a showstopper.

cc @fancyremarker @aaw

### CODE REVIEW

**Comment by aaw** [comment]
Looks like a good start!

A lot of database clients bypass SSL certificate verification by default (postgres, mysql, etc.). It looks like RabbitMQ is another example? Using a certificate signed by a trusted CA and passed in via `SSL_CERTIFICATE` and `SSL_KEY` is another layer of protection, so I'd say generally to use the `SSL_CERTIFICATE`/`SSL_KEY` mechanism if there's any chance that clients will attempt to verify it, otherwise using a self-signed certificate like you're doing here is okay.

If you want to use the `SSL_CERTIFICATE`/`SSL_KEY` mechanism, I don't know about the internals of RabbitMQ, but you may be able to either pass nothing as the cacertfile (which would mean use something in the local trust store) or else parse out the self-signed root from the certificate bundle (last certificate in the bundle) and try passing that as cacertfile. 

If you don't use the `SSL_CERTIFICATE`/`SSL_KEY` mechanism, I'd prefer that the self-signed certificate is generated during the database's `--initialize` call (like aptible/docker-mysql does) so that at least each database uses a different key pair and we're not using a single golden self-signed key pair that's committed to the repo.

The default options look fine to me, SSLv3 looks like it's disabled by default.

If the hostname needs to be stable, I don't see any problem with patching sweetness to force the hostname on every database container to be the database's Aptible handle. I can point you in the right direction of the sweetness changes needed if it isn't obvious.

**Comment by blakepettersson** [comment]
I went with your recommendation to generate the certificates on `--initialize` instead of going down the `SSL_CERTIFICATE/SSL_KEY` route. 

Apart from patching sweetness, everything else should be good to go unless if any of you guys has anything to add. 

cc @fancyremarker @aaw

File: wrapper.sh [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    rabbitmq-server &
+
+    # Capture the PID
+    rmq_pid=$!
+
+    # Tail the logs, but continue on to the wait command
+    echo -e "\n\nTailing log output:"
+
+    # If RMQ dies, this script dies
+    wait $rmq_pid
```
Comment (fancyremarker) [inline wrapper.sh]: :+1:

File: rabbitmqadmin [hunk 2]
```diff
@@ -0,0 +1,979 @@
+#!/usr/bin/env python
```
Comment (fancyremarker) [inline rabbitmqadmin]: Should this script be made accessible via the `--client` flag on wrapper.sh?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            tar -xzvf rabbitmq-server-generic-unix-${RABBITMQ_VERSION}.tar.gz && \
+            rm -f rabbitmq-server-generic-unix-${RABBITMQ_VERSION}.tar.gz && \
+            touch /srv/rabbitmq_server-${RABBITMQ_VERSION}/etc/rabbitmq/enabled_plugins && \
+            /srv/rabbitmq_server-${RABBITMQ_VERSION}/sbin/rabbitmq-plugins enable --offline rabbitmq_management && \
+            apk del --purge tar gzip && \
+            mkdir $DATA_DIRECTORY
+
+COPY        openssl.cnf /ssl/testca/openssl.cnf
+COPY        rabbitmq.config /srv/rabbitmq_server-${RABBITMQ_VERSION}/etc/rabbitmq/
+COPY        wrapper.sh /usr/bin/wrapper
```
Comment (fancyremarker) [inline Dockerfile]: It's not necessary for standardization but for consistency and memorability, it may be worth naming this `run-database.sh`.

**Comment by fancyremarker** [comment]
The structure of this looks very strong. One change worth making relates to the fact that we need to have a versioned tag for the Docker image, in case we release an upgraded and data-incompatible version of RabbitMQ in the future. So, I think we should:

1. Determine whether the version tag should be `3`, `3.5`, or `3.5.7`. Without doing any research, my gut predicts it is `3.5`, but there's a more exhaustive discussion of how we should decide on version tags on aptible/support#110.
2. To support multiple versions in the same repo, we probably want to move the Dockerfile and dependent files into a subdirectory, like we've done on docker-postgresql, docker-redis, etc.

I wasn't able to successfully build this locally (Bats tests were hanging) — are those still a work in progress?

**Comment by blakepettersson** [comment]
> I wasn't able to successfully build this locally (Bats tests were hanging) — are those still a work in progress?

The tests do work, but running the tests are slow due to starting and stopping RabbitMQ on each test. AFAIK there's nothing in bats that's equivalent to `setup_suite` or `teardown_suite`, which means that the time to run these tests is quite significant.

We could of course add a flag or something similar to `setup` and `teardown` to speed the tests up.

**Comment by fancyremarker** [comment]
Thanks for that clarification. I just tried building on my older Macbook, and didn't get the same build errors as on my newer Macbook at the office. (Perhaps it's a timing issue?) I'll retry the `docker build` on the computer where it failed before and respond back here with an actual build backtrace, if there is one.

**Comment by blakepettersson** [comment]
That's great, thanks! In the meantime I'll address your other comments.

**Comment by fancyremarker** [comment]
I may be a :doughnut: . I can't repro the build error I was seeing yesterday. Let's ignore that comment unless build errors appear on Quay or elsewhere.

**Comment by blakepettersson** [comment]
I'm looking into adding `--client` support now, I've fixed the other issues though as well as adding a Travis build for this project.

**Comment by blakepettersson** [comment]
I'm not sure how we would support `--client` for RabbitMQ; all other clients that we have open up a REPL-style client (`psql`, `redis-cli` etc) while rabbitmqadmin is little more than a glorified python script that fires off one-off commands.

Thoughts @fancyremarker?

**Comment by fancyremarker** [comment]
I vote no `--client` option in that case. We can just politely [print an error and exit 1](https://github.com/aptible/docker-elasticsearch/blob/ba71d7f/1.5/run-database.sh#L16-L17) if that option is called.

**Comment by fancyremarker** [comment]
@blakepettersson: Is this good for final review/merge? If not, what other changes remain?

**Comment by blakepettersson** [comment]
@fancyremarker for the docker image itself there should be nothing left. There is a pending PR for Sweetness, and beyond that there _should_ be just some minor fixes in dashboard.

**Comment by fancyremarker** [comment]
Merging this one; thanks for working through all the snags! I suspect we may want to refactor this to match the new repo structure of aptible/docker-redis#10 and aptible/docker-mysql#15, but that can happen separately.

**Pull request merged**